node/openrpc.go

```diff
@@ -0,0 +1,472 @@
+// Since we can't get the protocol scheme from the net.Listener itself, we have to define this for each
+// transport-specific document.
+func metaRegistererForURL(scheme string) *go_openrpc_reflect.MetaT {
+	var metaRegisterer = *sharedMetaRegisterer
+	metaRegisterer.GetServersFn = func() func(listeners []net.Listener) (*meta_schema.Servers, error) {
+		return func(listeners []net.Listener) (*meta_schema.Servers, error) {
+			servers := []meta_schema.ServerObject{}
+			for _, listener := range listeners {
+				url := scheme + listener.Addr().String()
+				network := listener.Addr().Network()
+				servers = append(servers, meta_schema.ServerObject{
+					Url:  (*meta_schema.ServerObjectUrl)(&url),
+					Name: (*meta_schema.ServerObjectName)(&network),
+				})
+			}
+			return (*meta_schema.Servers)(&servers), nil
+		}
+	}
+	return &metaRegisterer
+}
+
+// newOpenRPCDocument returns a Document configured with application-specific logic.
+func newOpenRPCDocument() *go_openrpc_reflect.Document {
+	d := &go_openrpc_reflect.Document{}
+
+	// Use a provided Ethereum default configuration as a base.
+	appReflector := &go_openrpc_reflect.EthereumReflectorT{}
+
+	// Install overrides for the json schema->type map fn used by the jsonschema reflect package.
+	appReflector.FnSchemaTypeMap = func() func(ty reflect.Type) *jsonschema.Type {
+		return OpenRPCJSONSchemaTypeMapper
+	}
+
+	appReflector.FnIsMethodEligible = func(method reflect.Method) bool {
+
+		// Method must be exported.
+		if method.PkgPath != "" {
+			return false
+		}
+
+		// Exclude methods that handle subscriptions, but do so without adhering to the conventional code pattern.
+		// Eg. *filters.PublicFiltersAPI.SubscribeNewHeads handles eth_subscribe("newHeads"), but there
+		// isn't a method called `eth_subscribeNewHeads`. So we blacklist all these methods and use
+		// the mock subscription receiver type RPCSubscription.
+		// This pattern matches all strings that start with Subscribe and are suffixed with a non-zero
+		// number of A-z characters.
+		if isPubSub(method.Type) {
+			// This catches MOST of them (except SubscribeSyncStatus)
+			return false
+		}
+		// Reject all methods that use a channel in the their argument params.
+		// This causes SubscribeSyncStatus to be rejected.
+		for i := 0; i < method.Func.Type().NumIn(); i++ {
+			in := method.Func.Type().In(i)
+			for in.Kind() == reflect.Ptr {
+				in = in.Elem()
+			}
+			if in.Kind() == reflect.Chan {
+				return false
+			}
+		}
+
+		// Verify return types. The function must return at most one error
+		// and/or one other non-error value.
+		outs := make([]reflect.Type, method.Func.Type().NumOut())
+		for i := 0; i < method.Func.Type().NumOut(); i++ {
+			outs[i] = method.Func.Type().Out(i)
+		}
+
+		// If an error is returned, it must be the last returned value.
+		switch {
+		case len(outs) > 2:
+			return false
+		case len(outs) == 1 && isErrorType(outs[0]):
```

Comment:
Is there an edge case that this case covers? In this case we might want to add a comment reasoning on it.

Can it be that we had in mind the opposite `!isErrorType(outs[0])`, though it doesn't make much sense as there are API methods like `web3_sha3` that return a non error output.

Reply:
Yea, it would be helpful to include examples in comments here. You can cross reference the `rpc` package to check corrrectness... in `rpc/server.go` I believe (?). The idea (as from the comment on #L134) is that `If an error is returned, it must be the last returned value`, which is just an enforced convention for the signature pattern for eligible methods.

- OK: `func a() error`
- OK: `func b() (int, error)`
- OK: `func c()`
- NOTOK: `func d() (error, int)`

Reply:
Resolved by https://github.com/etclabscore/core-geth/pull/220/commits/cfd5f28e7559180615b96c54a203ab144812f928

Reply:
Even better https://github.com/etclabscore/core-geth/pull/220/commits/690b2d392e2e57f49bc2403d93622f8b16440726. Thanks @meowsbits